### PR TITLE
Automated Release Process

### PR DESCRIPTION
**Pull request by jshcrowthe**
In an effort to ensure that we are more consistent and transparent w/ our release process, I have built a CLI tool that will automate the release pipeline so that anyone w/ rights to NPM should be able to release the SDK.

## Process Description

The SDK Release Process is as follows:

1. Check if there are packages to be updated
1. Check if the intended release is a staging or production release
1. Prompt the release engineer for the versions of each of the packages that are going to be updated with the current release
1. Validate the selected versions
1. Update the versions and the dependencies throughout the SDK
1. Clean the current git tree of any additional files (e.g. dependencies, built binaries, etc)
1. Reinstall Dependencies
1. Rebuild the SDK
1. Test the latest built SDK
1. Commit the changed versions (if this fails, the build will exit at this point)
1. Tag the latest commit with the appropriate version tags
1. Validate that the user is ready to push to GitHub/NPM
1. Push the changes to Github
1. Push the changes to NPM

## Publishing Flow (Diagrams help right?)
firebase js sdk - oss release flow - user flow

### CODE REVIEW

**Comment by merlinnot** [comment]
Don’t forget the bower, please :) Some of us are still forced to use it.

File: scripts/release/utils/constants.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { dirname, resolve } = require('path');
+exports.root = dirname(resolve(__dirname, '../../../package.json'));
```
Comment (hiranya911) [inline scripts/release/utils/constants.js]: Perhaps call this `projectRoot` for disambiguation

File: scripts/release/utils/npm.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+const { readFile: _readFile } = require('fs');
+const { promisify } = require('util');
+const Listr = require('listr');
+const readFile = promisify(_readFile);
+
+async function publishPackage(pkg, releaseType) {
+  try {
+    const path = await mapPkgNameToPkgPath(pkg);
+
+    const { private } = JSON.parse(
```
Comment (hiranya911) [inline scripts/release/utils/npm.js]: Recommend using a different variable name.

File: scripts/release/cli.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  try {
+    /**
+     * Welcome to the firebase release CLI!
+     */
+    await bannerText();
+
+    /**
+     * If there are unstaged changes, error
+     */
+    if (await hasDiff()) {
```
Comment (hiranya911) [inline scripts/release/cli.js]: Should probably check which branch this is on at this point.

File: .travis.yml [hunk 4]
```diff
@@ -29,3 +29,10 @@ addons:
     packages:
       - google-chrome-stable
       - g++-4.8
+
+# NPM Canary Build Config
+deploy:
+  - provider: script
+    script: echo "//registry.npmjs.org/:_authToken=\${NPM_TOKEN}" > .npmrc && yarn release --canary
```
[the remark this answers is not in the document]
Comment (jshcrowthe) [reply .travis.yml]: This is the format used for `.npmrc` files. I verified my local `.npmrc` includes this format and followed a tutorial [here](https://remysharp.com/2015/10/26/using-travis-with-private-npm-deps) indicating the same.

File: scripts/release/utils/inquirer.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+function capitalize(string) {
+  return string.charAt(0).toUpperCase() + string.slice(1);
+}
+
+exports.packageVersionUpdate = async (package, releaseType) => {
+  /**
+   * Get the current package version
+   */
+  let { version, private } = await mapPkgNameToPkgJson(package);
```
[the remark this answers is not in the document]
Comment (jshcrowthe) [reply scripts/release/utils/inquirer.js]: This is object dereferencing, though I could alias the property to something else, it doesn't really serve a purpose as `private` isn't a reserved word.

File: scripts/release/utils/npm.js (hunk 2 above)
Comment (jshcrowthe) [reply scripts/release/utils/npm.js]: See above about object dereferencing.

File: scripts/release/cli.js (hunk 3 above)
Comment (jshcrowthe) [reply scripts/release/cli.js]: Good call.

Comment (jshcrowthe) [reply scripts/release/cli.js]: Updated the comment above, WDYT?

File: scripts/release/utils/constants.js (hunk 1 above)
Comment (jshcrowthe) [reply scripts/release/utils/constants.js]: Done

File: scripts/release/cli.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    /**
+     * If there are no packages that have been updated
+     * skip the release cycle
+     */
+    if (!await hasUpdatedPackages()) {
+      console.log('No packages need to be updated. Exiting...');
+      return;
+    }
+
+    const whoami = await exec('npm whoami');
```
[the remark this answers is not in the document]
Comment (jshcrowthe) [reply scripts/release/cli.js]: Done

File: scripts/release/cli.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+       * Prompt for the release type (i.e. staging/prod)
+       */
+      const responses = await prompt([releaseTypePrompt]);
+      return responses.releaseType;
+    })();
+
+    let versions;
+
+    /**
+     * Set the canary version following, the pattern below:
```
[the remark this answers is not in the document]
Comment (jshcrowthe) [reply scripts/release/cli.js]: Done

File: scripts/release/utils/banner.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+const { readFile: _readFile } = require('fs');
+const { promisify } = require('util');
+const { resolve } = require('path');
+const readFile = promisify(_readFile);
+
+exports.bannerText = async () => {
+  let BANNER_TEXT = await readFile(resolve(__dirname, 'banner.txt'), 'utf8');
+  if (process.platform === 'darwin') {
```
Comment (mbleigh) [inline scripts/release/utils/banner.js]: FYI, the latest version of macOS breaks emoji spacing. We're likely removing the fire emoji soon :cry:

Comment (jshcrowthe) [reply scripts/release/utils/banner.js]: Sad dayyy... booooo

**Pull request merged**